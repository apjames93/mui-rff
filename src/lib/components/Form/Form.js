/* eslint-disable max-len */
import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import { Form as ReactFinalForm } from 'react-final-form';

export const RffWrapper = ({
  children,
  debug,
  initialValues,
  initialValuesEqual,
  keepDirtyOnReinitialize,
  onSubmit,
  subscription,
  validate,
  validateOnBlur,
  mutators,
}) => (
  <ReactFinalForm
    debug={debug}
    initialValues={initialValues}
    initialValuesEqual={initialValuesEqual}
    keepDirtyOnReinitialize={keepDirtyOnReinitialize}
    onSubmit={onSubmit}
    subscription={subscription}
    validate={validate}
    validateOnBlur={validateOnBlur}
    mutators={mutators}
  >
    {({ handleSubmit, submitting, form }) => (
      <form onSubmit={handleSubmit}>
        <>
          {Array.isArray(children) && children.map((child, key) => cloneElement(
            child,
            {
              key,
              submitting,
              form,
            },
          ))}

          {!Array.isArray(children) && cloneElement(
            children,
            {
              submitting,
              form,
            },
          )}
        </>
      </form>
    )}
  </ReactFinalForm>
);
export default RffWrapper;

// https://final-form.org/docs/react-final-form/types/FormProps
RffWrapper.propTypes = {
  /**
   * A render function that is given FormRenderProps, as well as any non-API props passed into the <Form/> component.
   */
  children: PropTypes.node.isRequired,
  /**
   * Function to call when the form is submitted. There are three possible ways to write an onSubmit function:
   */
  onSubmit: PropTypes.func.isRequired,

  /**
   * A callback for debugging that receives the form state and the states of all the fields. It's called on every state change. A typical thing to pass in might be console.log.
   */
  debug: PropTypes.func,
  /**
   * The initial values of your form. These will also be used to compare against the current values to calculate pristine and dirty.
   */
  initialValues: PropTypes.shape(),
  /**
   * A predicate to determine whether or not the initialValues prop has changed, i.e. to know if the form needs to be reinitialized with the new values. Useful for passing in a "deep equals" function if you need to. Defaults to "shallow equals".
   */
  initialValuesEqual: PropTypes.bool,
  /**
   * If true, only pristine values will be overwritten when initialize(newValues) is called. This can be useful for allowing a user to continue to edit a record while the record is being saved asynchronously, and the form is reinitialized to the saved values when the save is successful. Defaults to false.
   */
  keepDirtyOnReinitialize: PropTypes.bool,
  /**
   * An object of the parts of FormState to subscribe to. If a subscription is provided, the <Form/> will only rerender when those parts of form state change.
    If no subscription is provided, it will default to subscribing to all form state changes. i.e. <Form/> will rerender whenever any part of the form state changes.
   * ex : { [string]: boolean }
    */
  subscription: PropTypes.shape(),
  /**
   * ex: (values: FormValues) => Object | Promise<Object>
   * A whole-record validation function that takes all the values of the form and returns any validation errors. There are two possible ways to write a validate function:
   */
  validate: PropTypes.func,
  /**
   * If true, validation will happen on blur. If false, validation will happen on change. Defaults to false.
   */
  validateOnBlur: PropTypes.bool,
  /**
   * Optional.
   * Named mutator functions.
   */
  mutators: PropTypes.shape(),
};

RffWrapper.defaultProps = {
  debug: undefined,
  initialValues: {},
  initialValuesEqual: false,
  keepDirtyOnReinitialize: false,
  subscription: undefined,
  validate: undefined,
  validateOnBlur: false,
  mutators: undefined,
};
