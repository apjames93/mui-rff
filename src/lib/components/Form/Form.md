Form example:

```js noeditor
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';
```

```js
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';


const onSubmit = async (values) => {
  console.log('Input Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <MuiInput
      name="thing"
      label="thing"
    />
    <MuiInput
      name="number"
      label="number"
      type="number"
    />
    <MuiInput
      name="date"
      label="date"
      type="date"
    />
    <MuiSelect
      name="stuff"
      options={[{ num: 'six' }, { num: 'nine' }]}
      displayKey="num"
    />
    <MuiSubmit />
  </Form>
```
