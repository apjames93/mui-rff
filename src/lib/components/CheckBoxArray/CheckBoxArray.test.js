import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Form from '../Form/Form';

import {
  CheckBoxArray,
  handleOnChange,
  assignIndex,
  getIndex,
} from './CheckBoxArray';

function setup() {
  const props = {
    label: 'label',
    fields: {
      push: jest.fn(value => [value]),
      remove: jest.fn(),
      get: jest.fn(),
      forEach: jest.fn(),
      value: [{ id: 1 }],
    },
    value: { id: 2 },
    checked: false,
    color: 'primary',
  };
  const comp = shallow(<CheckBoxArray {...props} />);

  return { comp, props };
}

describe('onChange', () => {
  it.skip('onChange call handleOnChange', () => {
    const { props } = setup();
    const comp = mount(<CheckBoxArray {...props} />);
    jest.mock('./handleOnChange');

    const event = {
      preventDefault() {},
      target: { value: 'the-value' },
    };
    comp.find('#MUICheckbox').first().simulate('change', event);
    expect(handleOnChange).toBeCalled();
  });
});

describe('handleOnChange', () => {
  const value = { id: 1 };
  const fields = {
    push: jest.fn(),
    remove: jest.fn(),
  };

  it('handleOnChange push', () => {
    handleOnChange(value, fields, undefined);
    expect(fields.push).toBeCalledWith(value);
  });

  it('handleOnChange remove', () => {
    handleOnChange(value, fields, 1);
    expect(fields.remove).toBeCalledWith(1);
  });
});

describe('getIndex', () => {
  const fieldValue = { id: 1 };

  it('getIndex true', () => {
    const value = { id: 1 };
    const same = getIndex(fieldValue, value);
    expect(same).toBe(true);
  });

  it('getIndex false', () => {
    const value = { id: 2 };
    const same = getIndex(fieldValue, value);
    expect(same).toBe(false);
  });
});

describe('assignIndex', () => {
  const fields = {
    value: [{ id: 1 }, { id: 2 }],
  };

  it('assignIndex index', () => {
    const value = { id: 2 };
    const index = assignIndex(fields, value);
    expect(index).toBe(1);
  });

  it('getIndex false', () => {
    const value = { id: 5 };
    const index = getIndex(fields, value);
    expect(index).toBe(false);
  });
});

describe('<CheckBoxArray />', () => {
  it('renders CheckBoxArray', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create((
      <Form onSubmit={() => true}>
        <MemoryRouter>
          <CheckBoxArray {...props} />
        </MemoryRouter>
      </Form>

    ));
    expect(tree).toMatchSnapshot();
  });
});
