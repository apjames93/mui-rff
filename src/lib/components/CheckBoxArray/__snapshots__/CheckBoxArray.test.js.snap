// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBoxArray /> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    className="MuiFormControlLabel-root"
    field={
      Object {
        "input": Object {
          "checked": true,
          "name": undefined,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "type": "checkbox",
          "value": undefined,
        },
        "meta": Object {
          "active": false,
          "data": Object {},
          "dirty": true,
          "dirtySinceLastSubmit": false,
          "error": Object {},
          "initial": Object {},
          "invalid": true,
          "length": undefined,
          "modified": false,
          "modifiedSinceLastSubmit": false,
          "pristine": false,
          "submitError": undefined,
          "submitFailed": false,
          "submitSucceeded": false,
          "submitting": false,
          "touched": false,
          "valid": false,
          "validating": false,
          "visited": false,
        },
      }
    }
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          checked={false}
          className="PrivateSwitchBase-input-4"
          data-indeterminate={false}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
          value={
            Object {
              "id": 2,
            }
          }
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      label
    </span>
  </label>
</form>
`;
